### ISSUE TITLE
Question about multi-modal metric learning

### ISSUE THREAD

**Issue by gkouros** [KevinMusgrave/pytorch-metric-learning#305]
Hi Kevin,

I'm trying to implement a two-stream multi-modal metric learning pipeline, meaning that there needs to be a different trunk-embedder pair for each stream, however, I couldn't find any relevant example in the library. Is there one that I couldn't find? If not could you please provide some tips about how it would be best to implement this.

Thank you in advance!

George

**Comment by KevinMusgrave** [comment]
There isn't an existing solution. Here's something I put together quickly that might work (I haven't tested it):

```python
import torch

class WrappedModel(torch.nn.Module):
    def __init__(self, model1, model2):
    	self.model1 = model1
    	self.model2 = model2

    def forward(self, x, stream):
    	if stream == 1:
    		return self.model1(x)
    	elif stream == 2:
    		return self.model2(x)



class MultiModalTrainer(TwoStreamMetricLoss):
    def calculate_loss(self, curr_batch):
        (anchors, posnegs), labels = curr_batch
        embeddings = (
            self.compute_embeddings(anchors, stream=1),
            self.compute_embeddings(posnegs, stream=2),
        )

        indices_tuple = self.maybe_mine_embeddings(embeddings, labels)
        self.losses["metric_loss"] = self.maybe_get_metric_loss(
            embeddings, labels, indices_tuple
        )


    def compute_embeddings(self, data, **kwargs):
        trunk_output = self.get_trunk_output(data, **kwargs)
        embeddings = self.get_final_embeddings(trunk_output, **kwargs)
        return embeddings

    def get_final_embeddings(self, base_output, **kwargs):
        return self.models["embedder"](base_output, **kwargs)

    def get_trunk_output(self, data, **kwargs):
        data = c_f.to_device(data, device=self.data_device, dtype=self.dtype)
        return self.models["trunk"](data, **kwargs)



class MultiModalTester(GlobalTwoStreamEmbeddingSpaceTester):
    def compute_all_embeddings(self, dataloader, trunk_model, embedder_model):
        s, e = 0, 0
        with torch.no_grad():
            for i, data in enumerate(tqdm.tqdm(dataloader)):
                anchors, posnegs, label = self.data_and_label_getter(data)
                label = c_f.process_label(
                    label, self.label_hierarchy_level, self.label_mapper
                )
                a = self.get_embeddings_for_eval(trunk_model, embedder_model, anchors, stream=1)
                pns = self.get_embeddings_for_eval(trunk_model, embedder_model, posnegs, stream=2)
                if label.dim() == 1:
                    label = label.unsqueeze(1)
                if i == 0:
                    labels = torch.zeros(len(dataloader.dataset), label.size(1))
                    all_anchors = torch.zeros(len(dataloader.dataset), pns.size(1))
                    all_posnegs = torch.zeros(len(dataloader.dataset), pns.size(1))

                e = s + pns.size(0)
                all_anchors[s:e] = a
                all_posnegs[s:e] = pns
                labels[s:e] = label
                s = e
        return all_anchors, all_posnegs, labels


    def get_embeddings_for_eval(self, trunk_model, embedder_model, input_imgs, stream):
        input_imgs = c_f.to_device(
            input_imgs, device=self.data_device, dtype=self.dtype
        )
        trunk_output = trunk_model(input_imgs, stream)
        if self.use_trunk_output:
            return trunk_output
        return embedder_model(trunk_output, stream)
```

Use the ```MultiModalTrainer``` and ```MultiModalTester``` instead of the existing two-stream trainer and tester. 

Set the trunk and embedder like this:

```python
trunk = WrappedModel(trunk1, trunk2)
embedder = WrapperModel(embedder1, embedder2)
```

Explanation of above code: I just wrapped the 2 trunks and 2 embedders, and added a ```stream``` argument to the forward function. Then I had to add the ```stream``` argument in all the necessary places for the trainer and tester.

**Comment by PaulZgy** [comment]
Hi George,

I am facing the same problem now, have you found a solution? Can you give me some reference on how to realize it.

Thank you in advance!

Tom

**Comment by gkouros** [comment]
Hi Tom,
Kevin's suggestion worked fine for me, so I suggest starting from that.

**Comment by PaulZgy** [comment]
> Hi Tom,
> Kevin's suggestion worked fine for me, so I suggest starting from that.

thanks! I'm a beginner, so I have a lot of doubts, for example, how to deal with two different datasets, how to make two datasets select the same object, etc. If possible, can you send me your project for reference. 

Thank you very much!
